src/ir/fileReader.cpp

```diff
@@ -160,46 +197,64 @@ bool loadFromFile(Context* c, string filename,Module** top) {
       }
       if (jns.count("generators")) {
         for (auto jgenmap : jns.at("generators").get<jsonmap>()) {
-          string jgenname = jgenmap.first;
-          //TODO for now, if it has a module already, just skip
-          if (ns->hasGenerator(jgenname)) {
+          string genname = jgenmap.first;
+          if (ns->hasGenerator(genname)) {
             //TODO confirm that it has the same everything like genparams and modparams
             continue;
           }
 
           json jgen = jgenmap.second;
-          checkJson(jgen,{"typegen","genparams","defaultgenargs","metadata"});
+          checkJson(jgen,{"typegen","genparams"},{"modules","defaultgenargs","metadata"});
           Params genparams = json2Params(c,jgen.at("genparams"));
-          TypeGen* typegen = c->getTypeGen(jgen.at("typegen").get<string>());
-          assert(genparams == typegen->getParams());
-          Params modparams;
-          Generator* g = ns->newGeneratorDecl(jgenname,typegen,genparams);
+          
+          string typeGenName = jgen.at("typegen").get<string>();
+          ASSERTTHROW(c->hasTypeGen(typeGenName),"Missing typegen symbol " + typeGenName + " for generator " + jgenmap.first);
+          TypeGen* tg = c->getTypeGen(typeGenName);
+          vector<std::pair<Values,json>> genmodvalues;
+          //Verify that this is consistent with all the types
+          if (jgen.count("modules")) {
+            cout << "in modules!" << endl;
+            for (auto jgenmod : jgen.at("modules").get<jsonvector>()) {
+              jsonvector jvalmod = jgenmod.get<jsonvector>();
+              ASSERTTHROW(jvalmod.size()==2,"Bad generated module" + toString(jgenmod));
+              Values genargs = json2Values(c,jvalmod[0]);
+              json jmod = jvalmod[1];
+              checkJson(jmod,{"type"});
+              Type* t = json2Type(c,jmod.at("type"));
+              ASSERTTHROW(t==tg->getType(genargs),"Type mismatch for typegen\n  " + tg->toString() + toString(genargs) + "==" + t->toString() + " but != " +tg->getType(genargs)->toString());
+              genmodvalues.push_back({genargs,jmod});
+            }
+          }
+          //TODO deal with module parameter generation
+          Generator* g = ns->newGeneratorDecl(genname,tg,genparams);
           if (jgen.count("defaultgenargs")) {
             g->addDefaultGenArgs(json2Values(c,jgen.at("defaultgenargs")));
           }
           if (jgen.count("metadata")) {
             g->setMetaData(jgen["metadata"]);
           }
+          for (auto vjpair : genmodvalues) {
+            cout << "in loop:" << endl;
+            Module* m = g->getModule(vjpair.first);
+            m->print();
```

Comment:
Do we need these cout/print functions in stdout by default? Should consider adding some logging infrastructure so you can keep `stdout` clean for applications that don't require this information.